// @dcloudio/uni-uts-v1/dist/stacktrace/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUTSSyntaxError = exports.parseUTSRuntimeStacktrace = exports.parseRuntimeStacktrace = exports.parseCompileStacktrace = exports.parseUTSJavaScriptRuntimeStacktrace = exports.resolveUTSKotlinFilenameByClassName = exports.parseUTSKotlinRuntimeStacktrace = exports.parseUTSKotlinStacktrace = exports.parseUTSArkTSPluginStacktrace = exports.parseUTSSwiftPluginStacktrace = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const arkts_1 = require("./arkts");
const js_1 = require("./js");
const kotlin_1 = require("./kotlin");
const mp_1 = require("./mp");
const sourceMap_1 = require("../sourceMap");
const utils_1 = require("./utils");
const swift_1 = require("./swift");
var swift_2 = require("./swift");
Object.defineProperty(exports, "parseUTSSwiftPluginStacktrace", { enumerable: true, get: function () { return swift_2.parseUTSSwiftPluginStacktrace; } });
var arkts_2 = require("./arkts");
Object.defineProperty(exports, "parseUTSArkTSPluginStacktrace", { enumerable: true, get: function () { return arkts_2.parseUTSArkTSPluginStacktrace; } });
var kotlin_2 = require("./kotlin");
Object.defineProperty(exports, "parseUTSKotlinStacktrace", { enumerable: true, get: function () { return kotlin_2.parseUTSKotlinStacktrace; } });
Object.defineProperty(exports, "parseUTSKotlinRuntimeStacktrace", { enumerable: true, get: function () { return kotlin_2.parseUTSKotlinRuntimeStacktrace; } });
Object.defineProperty(exports, "resolveUTSKotlinFilenameByClassName", { enumerable: true, get: function () { return kotlin_2.resolveUTSKotlinFilenameByClassName; } });
var js_2 = require("./js");
Object.defineProperty(exports, "parseUTSJavaScriptRuntimeStacktrace", { enumerable: true, get: function () { return js_2.parseUTSJavaScriptRuntimeStacktrace; } });
function initEnv(options) {
    if (options.env) {
        if (options.env.UNI_COMPILER_VALIDATION_RULES_PATH) {
            process.env.UNI_COMPILER_VALIDATION_RULES_PATH =
                options.env.UNI_COMPILER_VALIDATION_RULES_PATH;
        }
    }
}
async function parseCompileStacktrace(stacktrace, options) {
    initEnv(options);
    if (options.platform === 'app-harmony' && options.language === 'arkts') {
        return (0, arkts_1.parseUTSArkTSPluginStacktrace)(stacktrace, options);
    }
    if (options.platform === 'app-ios' && options.language === 'swift') {
        return (0, swift_1.parseUTSSwiftPluginStacktrace)({
            ...options,
            stacktrace,
        });
    }
    return stacktrace;
}
exports.parseCompileStacktrace = parseCompileStacktrace;
async function parseRuntimeStacktrace(stacktrace, options) {
    initEnv(options);
    if ((options.platform === 'app-android' && options.language === 'kotlin') ||
        (options.platform === 'app-ios' && options.language === 'javascript') ||
        options.platform === 'app-harmony') {
        return parseUTSRuntimeStacktrace(stacktrace, options);
    }
    // mp-weixin:   sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析里边的base64格式sourceMap）
    // mp-baidu:    sourceMap无法合并映射（所以开发工具无法显示源码） 可以读取到sourceMap 二次解析映射（也可以合并sourceMap吧）
    // mp-toutiao:  sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析里边的base64格式sourceMap）
    // mp-alipay:   sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析url格式的sourceMap，再根据url下载sourceMap）
    if (mp_1.MP_PLATFORMS[options.platform]) {
        return (0, mp_1.parseMiniProgramRuntimeStacktrace)(stacktrace, options);
    }
    // 其他小程序平台暂不处理，因为没法拿到小程序的sourceMap做合并映射
    return stacktrace;
}
exports.parseRuntimeStacktrace = parseRuntimeStacktrace;
function parseUTSRuntimeStacktrace(stacktrace, options) {
    if (options.platform === 'app-harmony') {
        return (0, arkts_1.parseUTSHarmonyRuntimeStacktrace)(stacktrace, options);
    }
    else if (options.language === 'kotlin') {
        return (0, kotlin_1.parseUTSKotlinRuntimeStacktrace)(stacktrace, options);
    }
    else if (options.language === 'javascript') {
        return (0, js_1.parseUTSJavaScriptRuntimeStacktrace)(stacktrace, options);
    }
    return stacktrace;
}
exports.parseUTSRuntimeStacktrace = parseUTSRuntimeStacktrace;
function parseUTSSyntaxError(error, inputDir) {
    let errorMsg = error;
    if (error instanceof Error) {
        errorMsg = error.message;
        if (errorMsg.trim().startsWith('{')) {
            try {
                errorMsg = JSON.parse(errorMsg);
                return parseUTSSyntaxJsonError(errorMsg, inputDir);
            }
            catch (e) {
                return errorMsg;
            }
        }
    }
    return String(errorMsg).replace(/\t/g, ' ');
}
exports.parseUTSSyntaxError = parseUTSSyntaxError;
// {"message":"Expression expected","code":null,"level":"error","filename":"/Users/xxx/Documents/HBuilderProjects/test-vue3/unpackage/dist/dev/.uvue/app-android/uni_modules/test-uts/utssdk/index.uts","line":3,"column":4}
function parseUTSSyntaxJsonError(error, inputDir) {
    const normalizedError = new Error(error.message);
    const sourceMapFilename = error.filename + '.map';
    if (fs_extra_1.default.existsSync(sourceMapFilename)) {
        const result = (0, sourceMap_1.originalPositionForSync)({
            sourceMapFile: sourceMapFilename,
            line: error.line,
            column: error.column,
            withSourceContent: true,
        });
        if (result && result.source) {
            Object.defineProperty(normalizedError, 'id', {
                get() {
                    return path_1.default.resolve(inputDir, result.source);
                },
                set(_v) { },
            });
            normalizedError.loc = {
                file: result.source,
                line: result.line,
                column: result.column,
            };
            if (result.sourceContent) {
                normalizedError.frame = (0, utils_1.generateCodeFrame)(result.sourceContent, {
                    line: result.line,
                    column: result.column,
                }).replace(/\t/g, ' ');
            }
            return normalizedError;
        }
    }
    // 锁定id，防止rollup修改id
    Object.defineProperty(normalizedError, 'id', {
        get() {
            return error.filename;
        },
        set(_v) { },
    });
    // 解析 sourcemap
    normalizedError.loc = {
        file: error.filename,
        line: error.line,
        column: error.column,
    };
    normalizedError.frame = error.frame || '';
    return normalizedError;
}
