// @dcloudio/uni-uts-v1/dist/arkts/compiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundleArkTS = void 0;
const path_1 = __importDefault(require("path"));
const utils_1 = require("./utils");
const utils_2 = require("../utils");
const stacktrace_1 = require("../stacktrace");
async function bundleArkTS({ isX, filename, rootDir, outDir, sourceMap, uni_modules, banner, footer, }) {
    const runtimePackageName = (0, utils_1.getRuntimePackageName)(isX);
    const buildOptions = {
        hbxVersion: process.env.HX_Version || process.env.UNI_COMPILER_VERSION,
        input: {
            root: rootDir,
            filename,
            paths: {
                '@dcloudio/uni-runtime': runtimePackageName,
            },
            uniModules: uni_modules,
            parseOptions: {
                tsx: true,
                noEarlyErrors: true,
                allowComplexUnionType: true,
                allowTsLitType: true,
            },
        },
        output: {
            errorFormat: 'json',
            outDir,
            package: '',
            banner,
            footer,
            imports: [],
            sourceMap: sourceMap ? path_1.default.resolve((0, utils_2.resolveUTSSourceMapPath)()) : false,
            extname: '.ets',
            logFilename: false,
            isPlugin: true,
            transform: {
                autoImportExternals: (0, utils_1.getArkTSAutoImports)(isX),
                uniExtApiDefaultNamespace: '@dcloudio/uni-app-x-runtime',
            },
            treeshake: {
                noSideEffects: true,
            },
        },
    };
    const { bundle, UTSTarget } = (0, utils_2.getUTSCompiler)();
    const result = await bundle(UTSTarget.ARKTS, buildOptions);
    if (!result) {
        return;
    }
    if (result.error) {
        throw (0, stacktrace_1.parseUTSSyntaxError)(result.error, process.env.UNI_INPUT_DIR);
    }
    (0, utils_2.normalizeUTSResult)('app-harmony', result);
    const deps = [filename];
    if (process.env.NODE_ENV === 'development') {
        if (result.deps) {
            deps.push(...result.deps);
        }
    }
    if (result.inject_apis && result.inject_apis.length) {
        (0, utils_2.addPluginInjectApis)(result.inject_apis);
    }
    return result;
}
exports.bundleArkTS = bundleArkTS;
