// @dcloudio/uni-uts-v1/dist/arkts/encrypt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileEncrypt = void 0;
const path_1 = __importStar(require("path"));
const shared_1 = require("../shared");
const utils_1 = require("../utils");
async function compileEncrypt(pluginDir, _isX = false) {
    return compileEncryptByUniHelpers(pluginDir);
}
exports.compileEncrypt = compileEncrypt;
async function compileEncryptByUniHelpers(pluginDir) {
    const inputDir = process.env.UNI_INPUT_DIR;
    const outputDir = process.env.UNI_OUTPUT_DIR;
    const pluginId = path_1.default.basename(pluginDir);
    const pluginRelativeDir = (0, path_1.relative)(inputDir, pluginDir);
    const outputPluginDir = (0, shared_1.normalizePath)((0, path_1.join)(outputDir, pluginRelativeDir));
    const cachePluginDir = path_1.default.resolve(process.env.UNI_MODULES_ENCRYPT_CACHE_DIR, pluginRelativeDir);
    const encryptHarFile = (0, path_1.join)(cachePluginDir, 'module.har');
    const { DUMB } = (0, utils_1.requireUniHelpers)();
    try {
        await DUMB(path_1.default.basename(pluginRelativeDir), encryptHarFile, (0, path_1.join)(outputPluginDir, 'utssdk', 'app-harmony', 'module.har'));
    }
    catch (e) {
        console.error(e);
        process.exit(0);
    }
    return {
        dir: outputPluginDir,
        code: (0, utils_1.requireUTSPluginCode)(pluginId, false),
        deps: [],
        encrypt: true,
        inject_apis: [],
        scoped_slots: [],
        custom_elements: {},
    };
}
