// @dcloudio/uni-uts-v1/dist/arkts/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveAppHarmonyUniModulesEntryDir = exports.resolveAppHarmonyUniModuleDir = exports.resolveAppHarmonyUniModulesRootDir = exports.compileArkTS = exports.compileArkTSExtApi = exports.bundleArkTS = exports.getArkTSAutoImports = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const utils_1 = require("../utils");
const fast_glob_1 = require("fast-glob");
const shared_1 = require("../shared");
const stacktrace_1 = require("../stacktrace");
const utils_2 = require("./utils");
const encrypt_1 = require("./encrypt");
const encrypt_2 = require("../encrypt");
var utils_3 = require("./utils");
Object.defineProperty(exports, "getArkTSAutoImports", { enumerable: true, get: function () { return utils_3.getArkTSAutoImports; } });
var compiler_1 = require("./compiler");
Object.defineProperty(exports, "bundleArkTS", { enumerable: true, get: function () { return compiler_1.bundleArkTS; } });
/**
 * 将config.json内的依赖相对路径转为oh-package.json5的依赖相对路径
 */
function parsePackageDeps(deps) {
    if (!deps) {
        return;
    }
    const base = '/';
    const result = {};
    for (const key in deps) {
        const value = deps[key];
        if (value.startsWith('file:.') || value.startsWith('.')) {
            const configRelativePath = value.replace(/^file:/, '');
            const packageRelativePath = path_1.default
                .relative(base, path_1.default.resolve(base, 'utssdk/app-harmony', configRelativePath))
                .replace(/\\/g, '/');
            result[key] = packageRelativePath.startsWith('.')
                ? packageRelativePath
                : './' + packageRelativePath;
        }
        else {
            result[key] = value;
        }
    }
    return result;
}
async function compileArkTSExtApi(rootDir, pluginDir, outputDir, { isExtApi, isX, isOhpmPackage = false, sourceMap, uni_modules, rewriteConsoleExpr, }) {
    let filename = resolveAppHarmonyIndexFile(pluginDir);
    if (!filename) {
        // 如果有自定义组件，则使用自定义组件生成的index.uts
        const customElements = (0, utils_1.resolveCustomElements)(pluginDir);
        if (Object.keys(customElements).length === 0) {
            return;
        }
        filename = path_1.default.resolve(pluginDir, 'utssdk/app-harmony/index.uts');
    }
    const runtimePackageName = (0, utils_2.getRuntimePackageName)(isX);
    const { bundle, UTSTarget } = (0, utils_1.getUTSCompiler)();
    const pluginId = path_1.default.basename(pluginDir);
    const outputUniModuleDir = outputDir;
    let autoImportExternals = (0, utils_2.getArkTSAutoImports)(isX);
    if (isOhpmPackage) {
        // 只保留uni-app-runtime
        autoImportExternals = {
            [runtimePackageName]: autoImportExternals[runtimePackageName],
        };
    }
    const buildOptions = {
        hbxVersion: process.env.HX_Version || process.env.UNI_COMPILER_VERSION,
        input: {
            root: rootDir,
            filename: (0, utils_1.resolveBundleInputFileName)('app-harmony', filename),
            paths: {
                '@dcloudio/uni-runtime': runtimePackageName,
            },
            uniModules: uni_modules,
            parseOptions: {
                tsx: true,
                noEarlyErrors: true,
                allowComplexUnionType: true,
                allowTsLitType: true,
            },
        },
        output: {
            errorFormat: 'json',
            outDir: outputUniModuleDir,
            outFilename: 'utssdk/app-harmony/index.ets',
            package: '',
            imports: [],
            sourceMap: sourceMap
                ? path_1.default.resolve((0, utils_1.resolveUTSSourceMapPath)(), 'uni_modules', pluginId)
                : false,
            extname: '.ets',
            logFilename: false,
            isPlugin: true,
            transform: {
                autoImportExternals,
                uniExtApiDefaultNamespace: '@dcloudio/uni-app-x-runtime',
            },
            treeshake: {
                noSideEffects: true,
            },
        },
    };
    const configFilePath = path_1.default.resolve(pluginDir, 'utssdk/app-harmony/config.json');
    const harmonyPackageName = '@uni_modules/' + pluginId.toLowerCase();
    const harmonyModuleName = harmonyPackageName
        .replace(/@/g, '')
        .replace(/\//g, '__')
        .replace(/-/g, '_');
    // 拷贝所有ets、har文件
    const etsFiles = (0, fast_glob_1.sync)('**/*.{ets,js,har,tgz}', {
        cwd: pluginDir,
    });
    const depEtsFiles = [];
    for (const etsFile of etsFiles) {
        const srcFile = path_1.default.resolve(pluginDir, etsFile);
        const destFile = path_1.default.resolve(outputUniModuleDir, etsFile);
        if (/\.(ets|js)$/.test(etsFile)) {
            depEtsFiles.push(srcFile);
            if (rewriteConsoleExpr) {
                const content = fs_extra_1.default.readFileSync(srcFile, 'utf8');
                const newContent = rewriteConsoleExpr(srcFile, content);
                fs_extra_1.default.outputFileSync(destFile, newContent);
            }
            else {
                fs_extra_1.default.copySync(srcFile, destFile);
            }
        }
        else {
            fs_extra_1.default.copySync(srcFile, destFile);
        }
    }
    // generate oh-package.json5
    let version = '1.0.0';
    const packageJsonPath = path_1.default.resolve(pluginDir, 'package.json');
    if (fs_extra_1.default.existsSync(packageJsonPath)) {
        const packageJson = fs_extra_1.default.readJSONSync(packageJsonPath);
        version = packageJson.version || '1.0.0';
    }
    const ohPackageJson = {
        name: harmonyPackageName,
        version,
        description: '',
        main: 'utssdk/app-harmony/index.ets',
        author: '',
        license: '',
        dependencies: (uni_modules || []).reduce((acc, dep) => {
            acc['@uni_modules/' + dep.toLowerCase()] = '../' + dep;
            return acc;
        }, {}),
    };
    if (isOhpmPackage) {
        ohPackageJson.description = 'uni-app runtime package';
        ohPackageJson.author = 'DCloud';
        ohPackageJson.license = 'Apache-2.0';
    }
    if (fs_extra_1.default.existsSync(configFilePath)) {
        const config = fs_extra_1.default.readJSONSync(configFilePath);
        ohPackageJson.dependencies = parsePackageDeps(config.dependencies);
        ohPackageJson.dynamicDependencies = parsePackageDeps(config.dynamicDependencies);
        ohPackageJson.devDependencies = parsePackageDeps(config.devDependencies);
        ohPackageJson.overrides = parsePackageDeps(config.overrides);
    }
    fs_extra_1.default.outputJSONSync(path_1.default.resolve(outputUniModuleDir, 'oh-package.json5'), ohPackageJson, {
        spaces: 2,
    });
    // copy resources
    const resourcesDir = path_1.default.resolve(pluginDir, 'utssdk/app-harmony/resources');
    if (fs_extra_1.default.existsSync(resourcesDir)) {
        fs_extra_1.default.copySync(resourcesDir, path_1.default.resolve(outputUniModuleDir, 'src/main/resources'));
    }
    // TODO 以下文件用户可定制
    // src/main/module.json5
    const moduleJson5Path = path_1.default.resolve(pluginDir, 'utssdk/app-harmony/module.json5');
    const defaultModuleJson5Module = {
        name: harmonyModuleName,
        type: 'har',
        deviceTypes: ['default', 'tablet', '2in1'],
    };
    if (fs_extra_1.default.existsSync(moduleJson5Path)) {
        // merge module.json5
        const moduleJson5 = (0, shared_1.parseJson)(fs_extra_1.default.readFileSync(moduleJson5Path).toString('utf8'));
        if (!moduleJson5.module) {
            moduleJson5.module = defaultModuleJson5Module;
        }
        moduleJson5.module = Object.assign({}, defaultModuleJson5Module, moduleJson5.module);
        fs_extra_1.default.outputJSONSync(path_1.default.resolve(outputUniModuleDir, 'src/main/module.json5'), moduleJson5, {
            spaces: 2,
        });
    }
    else {
        fs_extra_1.default.outputJSONSync(path_1.default.resolve(outputUniModuleDir, 'src/main/module.json5'), {
            module: defaultModuleJson5Module,
        }, {
            spaces: 2,
        });
    }
    // build-profile.json5
    fs_extra_1.default.outputJSONSync(path_1.default.resolve(outputUniModuleDir, 'build-profile.json5'), {
        apiType: 'stageMode',
        buildOption: {},
        buildOptionSet: [],
        targets: [
            {
                name: 'default',
                config: {
                    buildOption: {
                        arkOptions: {
                            runtimeOnly: {
                                packages: [runtimePackageName],
                            },
                        },
                    },
                },
            },
        ],
    }, {
        spaces: 2,
    });
    // hvigorfile.ts
    fs_extra_1.default.outputFileSync(path_1.default.resolve(outputUniModuleDir, 'hvigorfile.ts'), `import { harTasks } from '@ohos/hvigor-ohos-plugin';

export default {
    system: harTasks,  /* Built-in plugin of Hvigor. It cannot be modified. */
    plugins:[]         /* Custom plugin to extend the functionality of Hvigor. */
}`);
    const result = await bundle(UTSTarget.ARKTS, buildOptions);
    if (!result) {
        return;
    }
    if (result.error) {
        throw (0, stacktrace_1.parseUTSSyntaxError)(result.error, process.env.UNI_INPUT_DIR);
    }
    (0, utils_1.normalizeUTSResult)('app-harmony', result);
    const deps = [filename, ...depEtsFiles];
    if (process.env.NODE_ENV === 'development') {
        if (result.deps) {
            deps.push(...result.deps);
        }
    }
    if (result.inject_apis && result.inject_apis.length) {
        (0, utils_1.addPluginInjectApis)(result.inject_apis);
    }
    return {
        code: (0, utils_1.requireUTSPluginCode)(pluginId, !!isExtApi),
        deps,
        encrypt: true,
        dir: outputUniModuleDir,
        inject_apis: [],
        scoped_slots: [],
        custom_elements: {},
    };
}
exports.compileArkTSExtApi = compileArkTSExtApi;
async function compileArkTS(pluginDir, options) {
    // 加密插件
    if ((0, encrypt_2.isEncrypt)(pluginDir)) {
        return (0, encrypt_1.compileEncrypt)(pluginDir, options.isX);
    }
    const inputDir = process.env.UNI_INPUT_DIR;
    const pluginId = path_1.default.basename(pluginDir);
    return compileArkTSExtApi((0, utils_1.resolveBundleInputRoot)('app-harmony', inputDir), pluginDir, resolveAppHarmonyUniModuleDir(pluginId), options);
}
exports.compileArkTS = compileArkTS;
function resolveAppHarmonyIndexFile(pluginDir) {
    let indexFile = path_1.default.resolve(pluginDir, 'utssdk/app-harmony/index.uts');
    if (fs_extra_1.default.existsSync(indexFile)) {
        return indexFile;
    }
    indexFile = path_1.default.resolve(pluginDir, 'utssdk/index.uts');
    if (fs_extra_1.default.existsSync(indexFile)) {
        return indexFile;
    }
}
function resolveAppHarmonyUniModulesRootDir() {
    if (process.env.UNI_APP_HARMONY_PROJECT_PATH) {
        return path_1.default.resolve(process.env.UNI_APP_HARMONY_PROJECT_PATH, 'uni_modules');
    }
    return path_1.default.resolve(process.env.UNI_OUTPUT_DIR, 'uni_modules');
}
exports.resolveAppHarmonyUniModulesRootDir = resolveAppHarmonyUniModulesRootDir;
function resolveAppHarmonyUniModuleDir(pluginId) {
    return path_1.default.resolve(resolveAppHarmonyUniModulesRootDir(), pluginId);
}
exports.resolveAppHarmonyUniModuleDir = resolveAppHarmonyUniModuleDir;
function resolveAppHarmonyUniModulesEntryDir() {
    if (process.env.UNI_APP_HARMONY_PROJECT_PATH) {
        return path_1.default.resolve(process.env.UNI_APP_HARMONY_PROJECT_PATH, 'entry/src/main/ets/uni_modules');
    }
    return path_1.default.resolve(process.env.UNI_OUTPUT_DIR, 'uni_modules');
}
exports.resolveAppHarmonyUniModulesEntryDir = resolveAppHarmonyUniModulesEntryDir;
